homeassistant/components/camera/synology.py

```diff
@@ -0,0 +1,131 @@
+"""
+Support for Synology Surveillance Station Cameras.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/camera.synology/
+"""
+import logging
+
+import voluptuous as vol
+
+import requests
+
+from homeassistant.const import (
+    CONF_NAME, CONF_USERNAME, CONF_PASSWORD,
+    CONF_URL, CONF_CAMERA_NAME, CONF_STREAM_ID)
+from homeassistant.components.camera import (Camera, PLATFORM_SCHEMA)
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = 'Synology Camera'
+DEFAULT_STREAM_ID = '0'
+TIMEOUT = 10
+
+# pylint: disable=line-too-long
+MJPEG_URL = '{0}/webapi/SurveillanceStation/streaming.cgi?api=\
+SYNO.SurveillanceStation.Streaming&method=LiveStream&version=1&cameraId={1}'
+STILL_IMAGE_URL = '{0}{1}'
+LOGIN_URL = '{0}/webapi/auth.cgi?api=SYNO.API.Auth&method=\
+Login&version=2&account={1}&passwd={2}&session=SurveillanceStation&format=sid'
+URL = '{0}/webapi/entry.cgi?\
+api=SYNO.SurveillanceStation.Camera&method=List&version=1'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Required(CONF_URL): cv.string,
+    vol.Required(CONF_CAMERA_NAME): cv.string,
+    vol.Optional(CONF_STREAM_ID, default=DEFAULT_STREAM_ID): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup a Synology IP Camera."""
+    add_devices([SynologyCamera(config)])
+
+
+# pylint: disable=too-many-instance-attributes
+class SynologyCamera(Camera):
+    """An implementation of a Synology NAS based IP camera."""
+
+    def __init__(self, config):
+        """Initialize a Synology Surveillance Station camera."""
+        from json import loads
+        super().__init__()
+        self._name = config.get(CONF_NAME)
+        self._username = config.get(CONF_USERNAME)
+        self._password = config.get(CONF_PASSWORD)
+        self._synology_url = config.get(CONF_URL)
+        self._camera_name = config.get(CONF_CAMERA_NAME)
+        self._stream_id = config.get(CONF_STREAM_ID)
+
+        self.get_sid()
+
+        # With our session id (sid) we can get the snapshot path.
+        sidurl = URL.format(self._synology_url)
+        req = requests.get(
+            sidurl, timeout=TIMEOUT, verify=False, cookies={'id': self._sid})
+        cam_resp = loads(req.text)
+        cameras = cam_resp['data']['cameras']
+
+        for camera in cameras:

```

Comment:
I moved this logic to  as requested.  Since we can't check on cameras without being authenticated, I also had to move authentication to the setup_platforms() method as well.